Use WordPress `CardBody` and `CardDivider` components for standard padding in deposits card

haszari opened the pull request:
Fixes #5954 

**Merge note**: This PR adjusts markup on two cards under heavy development (lots of open PRs). We've decided to hold off merging this PR until the current open PRs are merged, to avoid conflicts delaying the core feature work.

#### Changes proposed in this Pull Request

This PR refactors our two new cards (`DepositsOverview` and `DepositsInformationOverview` (aka Balances card?) to make full use of [card-related components from WordPress core](https://developer.wordpress.org/block-editor/reference-guides/components/card/). 

Previously, our card child components were individually adding the card padding.

Here's a screenshot of the new `DepositsOverview` card from this branch:

![Screenshot 2023-04-04 at 12 38 43 PM](https://user-images.githubusercontent.com/4167300/229656726-ae5175ce-a69a-4bba-9f76-5795b8cba560.png)

Specific changes:

- `DepositsInformationOverview` now uses `CardBody`. No other changes – this didn't break the existing styling. 
- `NextDepositDetails`
  - A `CardBody` for each child component (header, table header, table row). 
  - A `CardDivider` for the table header divider.
  - Removed some padding and custom divider style rules (now supplied by WP core styles).
- `DepositsOverview` has a class applied to the outer container `Card`. This is so can override (reset) `CardBody` padding as needed in the stacked `CardBody`s.

#### Testing instructions

* Set up a Woo store and onboard WCPay. Make some purchases so you have a balance to deposit.
* Enable the [deposits improvements feature flag](https://github.com/Automattic/woocommerce-payments/pull/5763): set `_wcpay_feature_simplify_deposits_ui` wp_option to `.
* Go to `Dashboard > Payments > Overview`.  
* Review the two deposits-related cards and ensure layout and look is correct and matches designs. 
* Test any interesting relevant deposits scenarios & ensure the cards are correct.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
consistently use CardBody container on deposits and balances cards

Commit:
remove placeholder footer, shift footer out of body

Commit:
remove extraneous horizontal padding from recent deposits table:
- now provided by CardBody

Commit:
use CardBody and CardDivider to standardise table styling

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5955` or branch name `fix/use-cardbody-component-for-standard-padding` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/use-cardbody-component-for-standard-padding) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 2377272c3ed4f1f71e032e8c606e7b27987500ba
- Build time: 2023-04-13 00:57:08 UTC

_Note: the build is updated when a new commit is pushed to this PR._

haszari commented on client/components/deposits-information/index.tsx:
```diff
@@ -106,31 +106,41 @@ const DepositsInformationOverview: React.FunctionComponent< OverviewProps > = (
-					children={ depositButton }
-				/>
-			</Flex>
+			<CardBody>
```
Using `CardBody` here for consistency. It didn't require any styling changes, but there may be redundant padding in styles for this component - reviewers please keep in mind!

haszari replied:
Side note: the name `DepositsInformationOverview` doesn't make sense to me, I think of this as the `Balances` card or maybe `CurrencySwitcher` card. 

Should we rename the main component? 

How would we refer to this UI in the documentation for merchants?

james-allan replied:
This `DepositsInformationOverview` is the old UI component. The one that looks like this: 

<img width="699" alt="Screenshot 2023-04-12 at 11 17 38 am" src="https://user-images.githubusercontent.com/8490476/231322384-670f457b-5892-4d82-9ffd-94e411f71005.png">

haszari replied:
Right … I missed that – thanks for clarifying. I'll remove these since we don't want to touch the existing UI code. 

Will investigate if changes are useful/necessary in the new balances card, let me know what you think :)

haszari replied:
These changes are now gone – thanks for the tip @james-allan! I reviewed `AccountBalances` components and it's looking good (no need to use `CardBody`, `CardDivider` etc).

james-allan replied:
Yep I was going to say the same. That card doesn't have any padding so we're all good there. :)

haszari committed:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

haszari commented on client/components/deposits-overview/index.tsx:
```diff
@@ -23,14 +23,11 @@ const DepositsOverview = (): JSX.Element => {
 			<p>Deposits History Section Goes here</p>
-
-			<p>Deposits Card Footer/Action Goes here</p>
-
```
I removed this while here, we have the footer implemented now! `<DepositsOverviewFooter />`

Jinksi commented:
Was the intention that this PR be merged before all existing project PRs, or that this PR "clean up" the styling of existing project PRs after they've been merged?

haszari commented:
> Was the intention that this PR be merged before all existing project PRs, or that this PR "clean up" the styling of existing project PRs after they've been merged?

Good question! I'm happy with either, though I think the idea (discussed in slack) was to merge other PRs relating to cards first, then update and merge this. I'll update the description to clarify.

haszari commented:
Marking as `blocked`: This PR adjusts markup on two cards under heavy development (lots of open PRs). We've decided to hold off merging this PR until the current open PRs are merged, to avoid conflicts delaying the core feature work.

haszari committed:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

Commit:
merge develop - adapting recent deposits history to CardBody (first cut):
- remove duplicate `Next deposit` heading
- wrap all of NextDeposit in CardBody – incl capital funds notice
- use CardBody and CardDivider for NextDeposit mini table
- SectionHeading is now a CardBody

Commit:
use table container class to override unnecessary CardBody vertical padding

Commit:
fix subheading colour for Next deposit

Commit:
use DepositOverviewSectionHeading for recent deposits heading for consistency, remove styling override

Commit:
shift "Next deposit" component to root (consistent with develop)

Commit:
remove unused import

haszari commented:
This is ready for review again – includes recent changes from `develop` including deposit history:

![Screenshot 2023-04-12 at 12 09 00 PM](https://user-images.githubusercontent.com/4167300/231315713-4b6caf66-2178-49a3-baa9-dc516e000d10.png)

haszari commented:
Removing `blocked` – lots of PRs have merged since this opened, only one other in review, and [deposit history is merged](https://github.com/Automattic/woocommerce-payments/pull/5850) (that was the biggest potential blocker / merge conflict).

Commit:
revert unnecessary CardBody changes in legacy/existing deposits info card

james-allan commented:
There's some extra padding being applied to the notices block. eg: 

<img width="696" alt="Screenshot 2023-04-12 at 11 38 46 am" src="https://user-images.githubusercontent.com/8490476/231325114-ed094e6d-9e4b-479d-86a3-5224e8c59174.png">

To resolve this, the `padding: 0 24px;` applied to the notices ([here](https://github.com/automattic/woocommerce-payments/blob/fix/use-cardbody-component-for-standard-padding/client/components/deposits-overview/style.scss#L26)) can be deleted.

Other than this this looks good.

Note that JS tests are failing because of the change in html outputted from these components. Running `npm run test:js -- -u` locally should resolve those out of date snapshots. Let me know if you'd like me to submit those updated snapshots. 

There's also a new conflict 😢. It was introduced as part of my [PR](https://github.com/Automattic/woocommerce-payments/pull/5839) so let me know if you would like me to resolve that.

james-allan reviewed (changes requested):
See https://github.com/Automattic/woocommerce-payments/pull/5955#issuecomment-1504418011

Commit:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

Commit:
remove unnecessary padding for notices – now handled by CardBody

Commit:
update js test snapshots

haszari commented:
Thanks @james-allan ! All that is fixed now:

> There's some extra padding being applied to the notices block

Removed the extra padding here a6826a73f9da690d655cdaa74ffd8492a54c59b2 – note I haven't tested all notices. 

> Note that JS tests are failing because of the change in html outputted from these components.

Aha! Thanks for the tip, now updated. 5d8bc1c85a6aa47be40a2088e87fec9c67561a58

> There's also a new conflict 😢

Sorted! 822af7e7f18c9921b4ce553dedbe081d28a87f92

Can you give this another quick test / review (double-check my merge)? If all's good feel free to kick off merge.

Commit:
add changelog placeholder

james-allan commented on client/components/deposits-overview/style.scss:
```diff
@@ -23,20 +23,27 @@
+
 	&__table {
 		color: $gray-900;
 		align-items: center;
```
```suggestion
		
```

These can be deleted as the table is now `__table__container`, these no longer apply and the lack of these two styles had no bearing on the actual result.

james-allan commented on client/components/deposits-overview/style.scss:
```diff
@@ -23,20 +23,27 @@
+	// so table CardDivider is full widh.
+	.components-card__body.wcpay-deposits-overview__table__container {
+		padding-bottom: 0;
+		padding-top: 0;
```
Because the table class was changed to `__table__container` the styles directly under the  `&__table` selector no longer apply.  

```suggestion
		padding-top: 0;
		.wcpay-banner-notice.components-notice {
			margin: 0;
			&:last-child {
				margin-bottom: 16px;
			}
		}
```

As I mentioned in the other review comment, the only affected styles were the direct children of the `__table` selector and this `wcpay-banner-notice` style because it wasn't an appending selector whereas the rest are.

Note that I've changed the `wcpay-banner-notice` styles above slightly because the margin left and right 24px are no longer needed.

haszari replied:
Thanks – now fixed f41d7b27f7f8ce7ec7831933d869639075160f85, I've moved that up to the table overrides section of the CSS and updated to suit.

Note I hacked the component to show the notice for testing:

<img width="830" alt="Screenshot 2023-04-12 at 4 26 50 PM" src="https://user-images.githubusercontent.com/4167300/231349663-f4e97e13-76ed-4692-beea-1858a50276fe.png">

james-allan reviewed (changes requested):
The latest changes look ok to me. Just two CSS issues I noticed in testing the deposit table notice. 

<img width="710" alt="Screenshot 2023-04-12 at 12 34 26 pm" src="https://user-images.githubusercontent.com/8490476/231346694-a686b2b5-54a2-44dd-9841-01185083ff4e.png">

Note the extra padding on the top and bottom of this notice.

haszari commented:
Failing tests look unrelated, anyone know if these are broken `develop`?

> Link with enabled UPE › Split UPE enabled › should save account details to Link
> Link with enabled UPE › Split UPE enabled › should use the saved Link account to fill in the details

haszari committed:
remove unused table styles

Co-authored-by: James Allan <james.allan@automattic.com>

james-allan commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -133,8 +139,12 @@ const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
+			</CardBody>
+			<CardDivider />
+			<CardBody className={ `${ tableClass }__container` }>
+				{ ' ' }
```
I also wanted to confirm that this is intentional?

haszari replied:
That is not intentional, I thought that was part of the merge! Investigating.

haszari replied:
Removed, bad berge 🫠 f41d7b27f7f8ce7ec7831933d869639075160f85

james-allan commented:
> Failing tests look unrelated, anyone know if these are broken develop?

Most likely related to this PR https://github.com/Automattic/woocommerce-payments/pull/5804. There's a matching revert PR too https://github.com/Automattic/woocommerce-payments/pull/6008

Commit:
override / remove margin on notice banners inside overview tables:
- e.g. "deposits may take 1-2 biz days" notice

Commit:
remove weird nullish merge artifact

haszari commented:
Should be good to go now @james-allan - thanks again :)

james-allan reviewed (approved):
Changes look good. 👍 

I've tested the previous scenarios:

- capital loan notice
- pending/in-transit deposit notice
- with and without a deposit history etc. 

The card looks as intended in all the scenarios I tested. 👍

james-allan commented:
> Changes look good.

Oh it looks like JS tests are failing again, most likely because of [78858eb](https://github.com/Automattic/woocommerce-payments/pull/5955/commits/78858eb60d852f988f6de8fbea0ece48f6407de9).

james-allan commented on client/components/deposits-overview/style.scss:
```diff
@@ -23,34 +23,40 @@
+		.wcpay-banner-notice.components-notice {
+			margin: 0;
+			&:last-child {
+				margin-bottom: 16px;
```
I might have given you some slightly bad advice regarding this. 

This margin bottom is being applied to the notices up in the next deposit section, when it was only intended to be applied to the notice in the deposit history table if its the last child. eg:

<img width="702" alt="Screenshot 2023-04-12 at 3 23 44 pm" src="https://user-images.githubusercontent.com/8490476/231358574-0bf08cdc-5293-4695-b33c-b15c79661603.png">

Notice the gap after the capital loan notice and the Deposit section heading.

It's supposed to look like this: 

<img width="700" alt="Screenshot 2023-04-12 at 3 25 07 pm" src="https://user-images.githubusercontent.com/8490476/231358837-c94a730c-52ef-4530-8c59-a508551e5dfd.png">

I don't have a good idea on how to resolve that. Either change this css selector to be more targeted so it only applies to the notices in the deposit history table. 

This is the scenario this style is meant to apply to. 
<img width="704" alt="Screenshot 2023-04-12 at 3 29 39 pm" src="https://user-images.githubusercontent.com/8490476/231359566-7ed9e1f3-51d3-4870-909d-71f627d4c146.png">

james-allan replied:
Looking over this style sheet there's a lot of instances (at least 3) of `margin: 0` being applied to the `.wcpay-banner-notice.components-notice` element. We might need to do a broader review of those notice styles tbh so let me know if this should be taken out of this issue.

haszari replied:
Thanks, I'll take a look at this today. I didn't test all the notices so thanks for catching :)

haszari replied:
I've hacked in all notices, and I _think_ it's looking good - though maybe there's a combination that's broken. I still need to test the "1 deposit which is in transit" case. 

![Screenshot 2023-04-13 at 10 28 10 AM](https://user-images.githubusercontent.com/4167300/231600260-8c2cda7f-392d-4fa9-bb74-0634f4ab6700.png)

james-allan replied:
The extra padding applies only when the Capital loan notice is shown. When there are other notices it looks right. 

<img width="692" alt="Screenshot 2023-04-13 at 9 50 35 am" src="https://user-images.githubusercontent.com/8490476/231610470-2f09a6b0-5322-4550-923f-4af99c28eca2.png">

The amount of padding between the notice and the **Deposit history** heading is an extra 16px. 

This is what it looks like on `develop`: 
<img width="702" alt="Screenshot 2023-04-13 at 9 52 58 am" src="https://user-images.githubusercontent.com/8490476/231610690-60459e6c-7d82-40b2-9a5d-d569758cbdc3.png">

The is happening because the capital loan notice is embedded in the `wcpay-deposits-overview__table__container` and so gets applied [this last child styling](https://github.com/automattic/woocommerce-payments/blob/fix/use-cardbody-component-for-standard-padding/client/components/deposits-overview/style.scss#L40-L42) whereas the deposit delay notice is in it's own div.

I think @brucealdridge has been looking at padding between these notices as part of https://github.com/Automattic/woocommerce-payments/issues/6002 so it might be handled as part of that.

haszari replied:
> The is happening because the capital loan notice is embedded in the wcpay-deposits-overview__table__container and so gets applied [this last child styling](https://github.com/automattic/woocommerce-payments/blob/fix/use-cardbody-component-for-standard-padding/client/components/deposits-overview/style.scss#L40-L42) whereas the deposit delay notice is in it's own div.

Maybe we can consistently put these two notices in same place in DOM – sounds like that would fix it.

haszari committed:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

haszari committed:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

Commit:
update test snapshot
